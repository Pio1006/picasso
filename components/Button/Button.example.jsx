import React from 'react'
import { storiesOf } from '@storybook/react'
import { text, select } from '@storybook/addon-knobs'
import IconSend from '@material-ui/icons/Send'

import StoryTeller from '../../.storybook/StoryTeller'
import Button from './'

const VARIANTS = {
  contained: 'contained',
  outlined: 'outlined',
  flat: 'flat'
}

const COLORS = {
  primary: 'primary',
  secondary: 'secondary',
  default: null
}

const stories = storiesOf('Button', module)

const teller = new StoryTeller('Button', 'A Button')
const chapter = teller.addChapter()

chapter
  .addSection('Contained', 'Contained buttons', () => (
    <div>
      <Button variant='contained'>Default</Button>
      <Button color='primary' variant='contained'>
        Primary
      </Button>
      <Button color='secondary' variant='contained'>
        Secondary
      </Button>
    </div>
  ))
  .addSection('Outlined', 'Outlined buttons', () => (
    <div>
      <Button variant='outlined'>Default</Button>
      <Button color='primary' variant='outlined'>
        Primary
      </Button>
      <Button color='secondary' variant='outlined'>
        Secondary
      </Button>
    </div>
  ))
  .addSection('Flat', 'Flat buttons', () => (
    <div>
      <Button>Default</Button>
      <Button color='primary'>Primary</Button>
      <Button color='secondary'>Secondary</Button>
    </div>
  ))
  .addSection(
    'Icon Button',
    "Buttons with icons (not supported yet as we don't have icon set",
    () => (
      <div>
        <Button icon={<IconSend />} variant='outlined'>
          Default
        </Button>
        <Button color='primary' icon={<IconSend />} variant='contained'>
          Primary
        </Button>
        <Button color='secondary' icon={<IconSend />}>
          Secondary
        </Button>
      </div>
    )
  )

stories.addWithChapters('Button', teller.toStory())

stories.add('variants', () => (
  <div className='section-component-container'>
    <Button
      color='primary'
      variant={select('variant', VARIANTS, VARIANTS.contained, 'MUI Props')}
    >
      {text('text', 'A Button')}
    </Button>
    <Button
      color='secondary'
      variant={select('variant', VARIANTS, VARIANTS.contained, 'MUI Props')}
    >
      {text('text', 'A Button')}
    </Button>
    <Button
      variant={select('variant', VARIANTS, VARIANTS.contained, 'MUI Props')}
    >
      {text('text', 'A Button')}
    </Button>
  </div>
))

stories.add('colors', () => (
  <div className='section-component-container'>
    <Button
      color={select('color', COLORS, COLORS.primary, 'MUI Props')}
      variant='outlined'
    >
      {text('text', 'A Button')}
    </Button>
    <Button
      color={select('color', COLORS, COLORS.primary, 'MUI Props')}
      variant='contained'
    >
      {text('text', 'A Button')}
    </Button>
    <Button color={select('color', COLORS, COLORS.primary, 'MUI Props')}>
      {text('text', 'A Button')}
    </Button>
  </div>
))
