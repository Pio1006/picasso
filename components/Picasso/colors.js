const colors = {
  primary: {
    main: '#204ecf',
    darker: '#1542c1'
  },
  secondary: {
    main: '#00cc83'
  },
  grey: {
    50: '#dfe3e9',
    100: '#686869'
  },
  blue: {
    light: '#cad5f4'
  },
  error: {
    main: '#9f3a38',
    text: '#e0b4b4',
    background: '#fff6f6'
  }
}

export default colors
