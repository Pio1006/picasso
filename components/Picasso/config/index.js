export { default as palette } from './palette'
export { default as transitions } from './transitions'
export { default as typography } from './typography'
