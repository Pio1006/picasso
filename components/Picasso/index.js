export { PicassoProvider } from './Picasso'
export { default } from './Picasso'
