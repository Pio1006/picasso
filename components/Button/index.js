import Button from './Button'
import { withPicasso } from '../theme'

export default withPicasso(Button)
