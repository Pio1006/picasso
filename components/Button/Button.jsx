import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import MUIButton from '@material-ui/core/Button'
import { withStyles } from '@material-ui/core/styles'
import { capitalize } from '@material-ui/core/utils/helpers'

import styles from './styles'
import Loader from '../Loader'
import Group from '../ButtonGroup'

const Button = props => {
  const {
    icon,
    loading,
    children,
    classes,
    variant,
    color,
    compact,
    ...rest
  } = props
  const {
    icon: iconClass,
    root: rootClass,
    children: childrenClass,
    hidden: hiddenClass,
    loader: loaderClass,
    ...restClasses
  } = classes
  let finalChildren = [children]

  if (icon) {
    const iconComponent = React.cloneElement(icon, { className: iconClass })

    finalChildren.unshift(iconComponent)
  }

  const rootClassName = cx(
    {
      [classes[`${variant.toLowerCase()}Negative`]]: color === 'negative',
      [classes.compact]: compact
    },
    classes[`${variant}${capitalize(color)}`],
    rootClass
  )

  return (
    <MUIButton
      classes={{
        root: rootClassName,
        ...restClasses
      }}
      color={color === 'negative' ? 'default' : color}
      variant={variant}
      {...rest}
    >
      <div className={cx(childrenClass, { [hiddenClass]: loading })}>
        {finalChildren}
      </div>

      {loading && <Loader className={loaderClass} inline size='small' />}
    </MUIButton>
  )
}

Button.propTypes = {
  color: PropTypes.string,
  compact: PropTypes.bool,
  icon: PropTypes.node,
  variant: PropTypes.string,
  loading: PropTypes.bool
}

Button.defaultProps = {
  color: 'default',
  compact: false,
  icon: null,
  loading: false,
  variant: 'outlined'
}

Button.Group = Group

export default withStyles(styles)(Button)
