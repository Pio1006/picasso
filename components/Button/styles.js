import { PicassoProvider } from '../Picasso'
import { RENDER_METHODS } from './utils/renderMethods'

const CONFIGURATION = ({ primary, secondary, grey, error, common }) => ({
  // contained variants
  contained: [RENDER_METHODS.contained, grey[50], common.black],
  containedPrimary: [RENDER_METHODS.contained, primary.main, common.white],
  containedSecondary: [RENDER_METHODS.contained, secondary.main, common.white],
  containedNegative: [RENDER_METHODS.contained, error.main, common.white],
  // outlined variants
  outlined: [RENDER_METHODS.outlined, grey[50], grey[100]],
  outlinedPrimary: [RENDER_METHODS.outlined, primary.main, primary.main],
  outlinedSecondary: [RENDER_METHODS.outlined, secondary.main, secondary.main],
  outlinedNegative: [RENDER_METHODS.outlined, error.main, error.main],
  // text (flat legacy) variants
  text: [RENDER_METHODS.text, grey[100], false],
  textPrimary: [RENDER_METHODS.text, primary.main, false],
  textSecondary: [RENDER_METHODS.text, secondary.main, false],
  textNegative: [RENDER_METHODS.text, error.main, false]
})

PicassoProvider.override(({ typography, transitions }) => ({
  MuiButton: {
    root: {
      textTransform: 'none',
      fontSize: typography.button.fontSize,
      padding: '0 1rem',
      transition: `border ${transitions.duration.short}ms ${
        transitions.easing.easeOut
      }, color ${transitions.duration.short}ms ${
        transitions.easing.easeOut
      }, background ${transitions.duration.short}ms ${
        transitions.easing.easeOut
      }`
    },
    label: {}
  }
}))

export default ({ palette }) => {
  const commonStyles = {
    icon: {
      fontSize: '1em',
      paddingRight: '0.45em'
    },
    compact: {
      minWidth: 'auto',
      '& $icon': {
        padding: 'initial'
      }
    }
  }

  const variants = CONFIGURATION(palette)

  for (const variant in variants) {
    const [renderMethod, primaryColor, secondaryColor] = variants[variant]

    commonStyles[variant] = renderMethod(primaryColor, secondaryColor)
  }

  return commonStyles
}
