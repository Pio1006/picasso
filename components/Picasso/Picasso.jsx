import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles'
import React from 'react'
import PT from 'prop-types'

import pallete from './pallete'
import Provider from './PicassoProvider'

const picasso = {
  pallete,
  props: {
    MuiButtonBase: {
      disableRipple: true
    }
  },
  shadows: Array.apply(null, Array(25)).map(() => 'none'), // Yest, this is needed as MUI expects 25 elements in array otherwise it raises error
  typography: {
    useNextVariants: true
  }
}

const createTheme = provider => createMuiTheme(provider.theme)

const Picasso = ({ children }) => (
  <MuiThemeProvider theme={createTheme(PicassoProvider)}>
    {children}
  </MuiThemeProvider>
)

Picasso.displayName = Picasso
Picasso.propTypes = {
  children: PT.node.isRequired
}

export const PicassoProvider = new Provider(picasso)
export default Picasso
