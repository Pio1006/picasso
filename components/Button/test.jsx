import React from 'react'
import { render, fireEvent, cleanup } from 'react-testing-library'

import Picasso from '../index'
import Button from './Button'

const renderButton = (children, props = {}) => {
  return render(
    <Picasso loadFonts={false}>
      <Button {...props}>{children}</Button>
    </Picasso>
  )
}

afterEach(cleanup)

test('onClick callback should be fired after clicking the button', () => {
  const onClick = jest.fn()
  const { getByText } = renderButton('Click me!', { onClick })

  fireEvent.click(getByText('Click me!'))

  expect(onClick).toHaveBeenCalled()
})

describe('disabled button', () => {
  let onClick
  let api

  beforeEach(() => {
    onClick = jest.fn()

    api = renderButton('Click me!', {
      onClick,
      disabled: true
    })
  })
  test('renders disabled version', () => {
    const { container } = api

    expect(container).toMatchSnapshot()
  })

  test('disables button events', () => {
    const { getByText } = api

    fireEvent.click(getByText('Click me!'))

    expect(onClick).not.toHaveBeenCalled()
  })
})
