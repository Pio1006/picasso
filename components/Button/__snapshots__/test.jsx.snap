// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contained button renders default contained button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-236 MuiButton-root-210 MuiButton-contained-221 MuiButton-raised-224"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-211"
    >
      Contained
    </span>
    <span
      class="MuiTouchRipple-root-239"
    />
  </button>
</div>
`;

exports[`contained button renders primary contained button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-273 MuiButton-root-247 MuiButton-contained-258 MuiButton-containedPrimary-259 MuiButton-raised-261 MuiButton-raisedPrimary-262"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-248"
    >
      Contained
    </span>
    <span
      class="MuiTouchRipple-root-276"
    />
  </button>
</div>
`;

exports[`contained button renders secondary contained button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-310 MuiButton-root-284 MuiButton-contained-295 MuiButton-containedSecondary-297 MuiButton-raised-298 MuiButton-raisedSecondary-300"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-285"
    >
      Contained
    </span>
    <span
      class="MuiTouchRipple-root-313"
    />
  </button>
</div>
`;

exports[`disabled button renders disabled version 1`] = `
<div>
  <button
    class="MuiButtonBase-root-65 MuiButtonBase-disabled-66 MuiButton-root-39 MuiButton-text-41 MuiButton-flat-44 MuiButton-disabled-59"
    disabled=""
    tabindex="-1"
    type="button"
  >
    <span
      class="MuiButton-label-40"
    >
      Click me!
    </span>
  </button>
</div>
`;

exports[`flat button renders default flat button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-347 MuiButton-root-321 MuiButton-text-323 MuiButton-flat-326"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-322"
    >
      Flat
    </span>
    <span
      class="MuiTouchRipple-root-350"
    />
  </button>
</div>
`;

exports[`flat button renders primary flat button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-384 MuiButton-root-358 MuiButton-text-360 MuiButton-textPrimary-361 MuiButton-flat-363 MuiButton-flatPrimary-364"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-359"
    >
      Flat
    </span>
    <span
      class="MuiTouchRipple-root-387"
    />
  </button>
</div>
`;

exports[`flat button renders secondary flat button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-421 MuiButton-root-395 MuiButton-text-397 MuiButton-textSecondary-399 MuiButton-flat-400 MuiButton-flatSecondary-402"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-396"
    >
      Flat
    </span>
    <span
      class="MuiTouchRipple-root-424"
    />
  </button>
</div>
`;

exports[`outlined button renders default outlined button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-125 MuiButton-root-99 MuiButton-outlined-107"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-100"
    >
      Outlined
    </span>
    <span
      class="MuiTouchRipple-root-128"
    />
  </button>
</div>
`;

exports[`outlined button renders primary outlined button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-162 MuiButton-root-136 MuiButton-outlined-144 MuiButton-outlinedPrimary-145"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-137"
    >
      Outlined
    </span>
    <span
      class="MuiTouchRipple-root-165"
    />
  </button>
</div>
`;

exports[`outlined button renders secondary outlined button 1`] = `
<div>
  <button
    class="MuiButtonBase-root-199 MuiButton-root-173 MuiButton-outlined-181 MuiButton-outlinedSecondary-183"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-174"
    >
      Outlined
    </span>
    <span
      class="MuiTouchRipple-root-202"
    />
  </button>
</div>
`;
